tutorials/tutorial-two-go-stream.md

```diff
@@ -0,0 +1,379 @@
+handlePublishConfirm(chPublishConfirm, messageCount, ch)
+
+fmt.Printf("Publishing %d messages\n", messageCount)
+for i := 0; i < messageCount; i++ {
+    var body string
+    if i == messageCount-1 {
+        body = "marker"
+    } else {
+        body = "hello"
+    }
+    producer.Send(amqp.NewMessage([]byte(body)))
+}
+_ = <-ch
+fmt.Println("Messages confirmed")
+
+producer.Close()
+```
+
+The sending program uses 2 channels and a Go routine to move on and exit when the broker confirmed the messages.
+Let's focus on this part.
+
+The producer `NotifyPublishConfirmation` function returns the first channel.
+The client library sends the broker confirmations on this channel and a routine declared in `handlePublishConfirm` receives these confirmations:
+
+```go
+messageCount := 100
+chPublishConfirm := producer.NotifyPublishConfirmation()
+ch := make(chan bool)
+handlePublishConfirm(chPublishConfirm, messageCount, ch)
+```
+
+The routine deals with the messages and sends `true` to the second channel when the expected number of confirmations is reached:
+
+```go
+func handlePublishConfirm(confirms stream.ChannelPublishConfirm, messageCount int, ch chan bool) {
+	go func() {
+		confirmedCount := 0
+		for confirmed := range confirms {
+			for _, _ = range confirmed {
+				confirmedCount++
+				if confirmedCount == messageCount {
+					ch <- true
+				}
+			}
+		}
+	}()
+}
+```
+
+The main program is waiting on the second channel just after the sending loop, so it moves on as soon as something arrives on the channel (the `true` value the routine sent).
+
+There is no risk the program stops before all the messages made it to the broker thanks to this synchronization mechanism.
+
+Let's now create the receiving program.
+
+### Receiving
+
+The receiving program creates the environment and declares the stream as well.
+This part of the code is the same as in the sending program, so it is skipped in the next code snippets for brevity's sake.
+
+The receiving program starts a consumer that attaches at the beginning of the stream (`stream.OffsetSpecification{}.First()`).
+It uses variables to output the offsets of the first and last received messages at the end of the program.
+
+The consumer stops when it receives the marker message: it assigns the offset to a variable, closes the consumer, and sends `true` to a channel.
+Like for the sender, the channel tells the program when to move on when the consumer is done with its job.
+
+```go
+var firstOffset int64 = -1
+var lastOffset atomic.Int64
+ch := make(chan bool)
+messagesHandler := func(consumerContext stream.ConsumerContext, message *amqp.Message) {
+    if atomic.CompareAndSwapInt64(&firstOffset, -1, consumerContext.Consumer.GetOffset()) {
+        fmt.Println("First message received.")
+    }
+    if string(message.GetData()) == "marker" {
+        lastOffset.Store(consumerContext.Consumer.GetOffset())
+        consumerContext.Consumer.Close()
```

Comment:
I usually suggest closing the consumer outside the message handler.

Reply:
done via: https://github.com/rabbitmq/rabbitmq-website/pull/1952/commits/81ddb9e4699a7701c026571ba99fc11e091c6837